src/ipyradiant/visualization/vis_tool.py

```diff
@@ -0,0 +1,114 @@
+from pathlib import Path
+
+import traitlets as T
+
+import ipywidgets as W
+from ipycytoscape import CytoscapeWidget
+from rdflib import Graph, Literal, URIRef
+from rdflib.namespace import RDF
+
+
+class CytoscapeVisualization(W.GridBox):
+    graph = T.Instance(Graph, allow_none=True)
+    cyto_widget = T.Instance(CytoscapeWidget, allow_none=True)
+    nodes = T.List()
+    click_output = T.Instance(W.Output)
+    click_output_box = T.Instance(W.VBox)
+    selected_node = T.Unicode(allow_none=True)
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.edge_color = kwargs.get("edge_color", "pink")
+        self.node_color = kwargs.get("node_color", "grey")
+        self.output_title = W.HTML("<h3>Output from Clicks</h3>")
+        self.click_output_box.children = [self.output_title, self.click_output]
+
+        self.default_edge = {"selector": "edge", "css": {"line-color": self.edge_color}}
+        self.default_node = {
+            "selector": "node",
+            "css": {"background-color": self.node_color},
+        }
+
+        self.children = [self.cyto_widget, self.click_output_box]
+
+    @T.default("click_output")
+    def _make_click_output(self):
+        return W.Output()
+
+    @T.default("nodes")
+    def _make_default_nodes(self):
+        return []
+
+    @T.default("click_output_box")
+    def _make_default_click_output_box(self):
+        return W.VBox()
+
+    @T.default("node_selector_box")
+    def _make_node_selector_box(self):
+        return W.VBox()
+
+    @T.default("cyto_widget")
+    def _make_default_cyto_widget(self):
+        cyto_widget = CytoscapeWidget()
+        cyto_widget.on("node", "click", self.log_node_clicks)
+        cyto_widget.on("edge", "click", self.log_edge_clicks)
+        cyto_widget.set_style([self.default_node, self.default_edge])
+        return cyto_widget
+
+    def log_node_clicks(self, node):
+        with self.click_output:
+            print(f"node clicked: {node['data']}")
+            print("-------------------------------")
+
+    def log_edge_clicks(self, edge):
+        with self.click_output:
+            print("edge clicked:")
+            print(f'edge source: {edge["data"]["source"]}')
+            print(f'edge target: {edge["data"]["target"]}')
+            print("-------------------------------")
+
+    @T.observe("graph")
+    def update_cyto_widget_graph(self, change):
+        # TODO configure vis tool with replace=True/False
```

Comment:
todos are good... new issues, and referencing them (e.g. `#99`) here, is better.

Reply:
You can reference this issue lukas
https://github.com/jupyrdf/ipyradiant/issues/15